[jest-worker] Detect infinite loops

ayshiff opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md at the root of the project if you have not done so. -->

## Summary

Closes #9785.

As said with @jeysal, here is the Draft PR to show where I am.
As I said [here](https://github.com/facebook/jest/issues/9785#issuecomment-636922330), I have been stuck for a moment on this and I will need some help 👍

I would like to have a TDD oriented approach to be able to make sure that the modifications I make are valid. Can someone give me any hints where to start to test the jest-worker heartbeat behavior as this is the part I'm having a hard time moving forward on ?

Currently the inspector initialization is done in `jest-haste-map` but I think it should be done in` jest-worker` or in `jest-runner`.

I made a graph to show the logic i tried to set up.

<p align="center">
<img width="600" src="https://user-images.githubusercontent.com/32459935/97105130-0dba3500-16b9-11eb-9885-70983df11f39.png" />
</p>


## Test plan

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

jeysal commented:
Alright, thanks for the work so far, it looks promising, let me try to respond to some things :)

> Currently the inspector initialization is done in `jest-haste-map` but I think it should be done in` jest-worker` or in `jest-runner`.

I agree with that `jest-haste-map` doesn't really have anything to do with this. I think `jest-worker` itself would be the best place since we can then also offer this functionality with a configurable timeout in its public interface to other consumers (it is actually a package used by many projects other than Jest!)

> Can someone give me any hints where to start to test the jest-worker heartbeat behavior as this is the part I'm having a hard time moving forward on ?

Hmm so I think this could be added to the existing tests of `processChild` and `ChildProcessWorker` (the latter would mock the `processChild` to emit the heartbeat messages). Later, once everything works, integration tests can be added.
A `processChild` test would probably use `jest.useFakeTimers` to check that the heartbeat is periodically sent.
Note that I think we will need to be able to pass the interval into the process. I think adding this number to the initialize message would make sense.

BTW, do you expect any difficulties with the Node thread workers or can they support the inspector in the same way that child process workers do?

SimenB commented on packages/jest-haste-map/src/index.ts:
```diff
@@ -711,17 +717,33 @@ class HasteMap extends EventEmitter {
 
+  private async setUpInspector() {
+    // Open V8 Inspector
+    await inspector.open();
```
neither this or `connect` return promises

SimenB commented on packages/jest-haste-map/src/index.ts:
```diff
@@ -711,17 +717,33 @@ class HasteMap extends EventEmitter {
+    if (inspectorUrl !== undefined) {
+      session = new inspector.Session();
+      await session.connect();
+      await session.post('Debugger.enable');
```
also doesn't return a promise, but you can promisify `post`

SimenB commented on packages/jest-haste-map/src/index.ts:
```diff
@@ -48,6 +48,8 @@ import type {
   WorkerMetadata,
 } from './types';
 
+import inspector = require('inspector');
```
```suggestion
import * as inspector from 'inspector';
```

I think

SimenB commented on packages/jest-haste-map/src/index.ts:
```diff
@@ -711,17 +717,33 @@ class HasteMap extends EventEmitter {
    * Creates workers or parses files and extracts metadata in-process.
    */
-  private _getWorker(options?: {forceInBand: boolean}): WorkerInterface {
+  private async _getWorker(options?: {forceInBand: boolean}): Promise<any> {
```
```suggestion
  private async _getWorker(options?: {forceInBand: boolean}): Promise<WorkerInterface> {
```

SimenB commented on packages/jest-worker/src/workers/ChildProcessWorker.ts:
```diff
@@ -8,6 +8,7 @@
 import {ChildProcess, fork} from 'child_process';
 import {PassThrough} from 'stream';
 import mergeStream = require('merge-stream');
+import {clearInterval} from 'timers';
```
why not use the global?

SimenB commented on packages/jest-worker/src/workers/ChildProcessWorker.ts:
```diff
@@ -157,13 +169,69 @@ export default class ChildProcessWorker implements WorkerInterface {
+      this._options.inspector.on('Debugger.paused', (message: any) => {
+        const callFrames = message.params.callFrames.slice(0, 20);
+        for (const callFrame of callFrames) {
+          const loc = callFrame['location'];
```
```suggestion
          const loc = callFrame.location;
```

etc

SimenB commented on packages/jest-worker/src/workers/ChildProcessWorker.ts:
```diff
@@ -157,13 +169,69 @@ export default class ChildProcessWorker implements WorkerInterface {
+      await new Promise((resolve, reject) => {
+        this._options.inspector.post(
+          'Debugger.pause',
+          (err: any, params: any) => {
```
```suggestion
          (err, params) => {
```

should be typed

SimenB commented:
(just left some nits on the current code, feel free to ignore while you figure out the approach to take 🙂)

ayshiff commented:
@jeysal @SimenB Thanks for your review 👍
I am still investigating about the proper way to implement this.

I moved the inspector initialization inside `jest-worker` and added  a configurable timeout in its public interface.
I also added the same logic as the ChildProcessWorker to the Node thread worker. I don't know yet if they can support the inspector the same way that child process workers do.

I keep testing the `heartbeat` logic between workers and parents but I am still a little lost in the way they communicate with each other.

Tell me if I am wrong but for me the `PARENT_MESSAGE_OK` indicates that the child process has succeed and it means that we can clear the heartbeat timeout.

jeysal commented:
I suppose there are multiple ways to go about this - either the child always sends heartbeats from when it is spawned until it is destroyed, or the child only sends heartbeats while working on a task. I suspect the former is easier, and the overhead of an interval sending one small message usually every couple of seconds seems negligible. So in that case the interval would be scheduled in initialize and cleared in `exitProcess`.

ayshiff commented:
> I suppose there are multiple ways to go about this - either the child always sends heartbeats from when it is spawned until it is destroyed, or the child only sends heartbeats while working on a task. I suspect the former is easier, and the overhead of an interval sending one small message usually every couple of seconds seems negligible. So in that case the interval would be scheduled in initialize and cleared in `exitProcess`.

You are right, it seems easier to handle !
I'll look in that direction then ! Thanks.

SimenB commented on packages/jest-worker/src/index.ts:
```diff
@@ -8,6 +8,8 @@
 
 import {cpus} from 'os';
+import * as inspector from 'inspector';
+import type {Session} from 'inspector';
```
```suggestion
```

just do `inspector.Session` where it's used

SimenB commented on packages/jest-worker/src/index.ts:
```diff
@@ -72,18 +74,23 @@ export default class JestWorker {
-      forkOptions: this._options.forkOptions ?? {},
-      maxRetries: this._options.maxRetries ?? 3,
-      numWorkers: this._options.numWorkers ?? Math.max(cpus().length - 1, 1),
+      enableWorkerThreads: this._options.enableWorkerThreads || false,
```
why the `??` to `||` changes?

ayshiff replied:
Oh.. I didn't pay enough attention while I was doing my rebase.

SimenB commented on packages/jest-worker/src/index.ts:
```diff
@@ -72,18 +74,23 @@ export default class JestWorker {
   private _farm: Farm;
   private _options: FarmOptions;
   private _workerPool: WorkerPoolInterface;
+  private _inspector: Session | undefined;
```
```suggestion
  private _inspectorSession: Session | undefined;
```

easier to grok when we have the `inspector` import

SimenB commented on packages/jest-worker/src/workers/ChildProcessWorker.ts:
```diff
@@ -66,6 +69,8 @@ export default class ChildProcessWorker implements WorkerInterface {
   private _exitPromise: Promise<void>;
   private _resolveExitPromise!: () => void;
 
+  private _heartbeatTimeout: any;
```
not `any`

SimenB commented on packages/jest-worker/src/workers/ChildProcessWorker.ts:
```diff
@@ -147,6 +153,13 @@ export default class ChildProcessWorker implements WorkerInterface {
+    clearTimeout(this._heartbeatTimeout);
+    this._heartbeatTimeout = setTimeout(() => {
+      onExceeded();
+    }, this._options.workerHeartbeatTimeout);
```
```suggestion
    }, this._options.workerHeartbeatTimeout).unref();
```

I think - we don't want this to keep the process open or some such

SimenB commented on packages/jest-worker/src/workers/ChildProcessWorker.ts:
```diff
@@ -147,6 +153,13 @@ export default class ChildProcessWorker implements WorkerInterface {
 
+  monitorHeartbeat(onExceeded: () => any): void {
+    clearTimeout(this._heartbeatTimeout);
+    this._heartbeatTimeout = setTimeout(() => {
```
why a `timeout` and not an `interval`?

ayshiff replied:
Maybe my logic is wrong then.

The child processes do `setInterval` to simulate the heartbeat behavior while the worker simply handles the case where there have been no messages from the child process for n-time.

From what I understand it may not be for child processes to fake the heartbeat but directly to the worker ?

![image](https://user-images.githubusercontent.com/32459935/97308583-501e7600-1861-11eb-899b-dfaf16b5c539.png)

jeysal replied:
No I think you're right, Simen probably just read over this very quickly :)

SimenB replied:
ah yes, ignore me 👍

SimenB commented on packages/jest-worker/src/workers/ChildProcessWorker.ts:
```diff
@@ -157,6 +170,58 @@ export default class ChildProcessWorker implements WorkerInterface {
+  private async monitorHeartbeatError() {
+    let error;
+
+    if (this._options.inspector !== undefined) {
```
flip `if-else` around so `true` case is first

ayshiff replied:
`true` case is already first, no ?

SimenB replied:
I mean check `===` first, not `!==`. Essentially https://eslint.org/docs/rules/no-negated-condition 🙂

SimenB commented on packages/jest-worker/src/workers/NodeThreadsWorker.ts:
```diff
@@ -42,6 +45,8 @@ export default class ExperimentalWorker implements WorkerInterface {
   private _resolveExitPromise!: () => void;
   private _forceExited: boolean;
 
+  private _heartbeatTimeout: any;
```
comments from child apply in this file as well (`any`, `timeout` vs `interval` etc)

SimenB commented on packages/jest-worker/src/workers/processChild.ts:
```diff
@@ -40,6 +43,10 @@ const messageListener: NodeJS.MessageListener = request => {
+      heartbeatIntervalValue = request[4];
+      monitorHeartbeat = setInterval(() => {
+        sendParentMessageHeartbeat();
+      }, heartbeatIntervalValue);
```
```suggestion
      }, heartbeatIntervalValue).unref();
```

here as well, I think?

SimenB commented on packages/jest-worker/src/workers/processChild.ts:
```diff
@@ -59,6 +66,12 @@ const messageListener: NodeJS.MessageListener = request => {
 process.on('message', messageListener);
 
+function sendParentMessageHeartbeat() {
+  if (process && process.send) {
```
```suggestion
  if (process?.send) {
```

when is `process` not defined?

SimenB commented on packages/jest-worker/src/workers/threadChild.ts:
```diff
@@ -42,6 +45,10 @@ const messageListener = (request: any) => {
       const init: ChildMessageInitialize = request;
       file = init[2];
       setupArgs = request[3];
+      heartbeatIntervalValue = request[4];
```
same comments as for process

SimenB commented:
again, feel free to ignore my nits, I was just stuck in line at the grocery store with nothing better to do 😀

ayshiff commented:
I try to test the current implementation with a simple test file containing an infinite loop to see how the workers and the child processes handle it:

```js
// infinite_loop.test.ts
it ('should handle infinite loop') {
  while (true) {}
}
```

But unfortunately when I try to test if the child processes emit correctly the `PARENT_MESSAGE_HEARTBEAT` to the parent within the `setInterval`, nothing is logged.

I don’t know if the logs are stopped when the child processes are working or if they don’t emit anything at all. (Maybe they get stopped before the first interval ends.)

SimenB commented on packages/jest-worker/src/index.ts:
```diff
@@ -105,6 +111,20 @@ export default class JestWorker {
+    if (inspectorUrl !== undefined) {
+      session = new inspector.Session();
+      session.connect();
+      session.post('Debugger.enable');
```
this is async

```suggestion
      await util.promisfy(session.post)('Debugger.enable');
```

ayshiff replied:
TS is complaining about the return type of `promisify`:   
`Argument of type '"Debugger.enable"' is not assignable to parameter of type '"NodeWorker.disable"'.`

It returns something with this type:
```js
// util.promisify(session.post)('Debugger.enable')
function util.promisify<"NodeWorker.disable", unknown>(fn: (arg1: "NodeWorker.disable", callback: (err: Error | null, result: unknown) => void) => void): (arg1: "NodeWorker.disable") => Promise<unknown> (+13 overloads)
```

Also, what is the best way to handle the asynchronous assignment of the inspector inside `constructor` ?

SimenB replied:
Ah right, I had issues with promisify myself https://github.com/facebook/jest/pull/8596#discussion_r296476038

Manual `new Promise` should work 👍

Regarding best way of async init is probably a setup function of some sort... Unless we can intercept the user's function calls as they are all async?

SimenB replied:
Or exposing a factory function instead of the class directly. Can do that in Jest 27 even if this PR doesn't land in time for it

SimenB commented on packages/jest-worker/src/workers/NodeThreadsWorker.ts:
```diff
@@ -137,6 +150,55 @@ export default class ExperimentalWorker implements WorkerInterface {
+      });
+    }
+    // @ts-expect-error: adding custom properties to errors.
+    error.type = 1;
```
`1`? should we add a `name`?

ayshiff replied:
Should we add another `PARENT_MESSAGE_ERROR` ?

```ts
// types.ts
export const PARENT_MESSAGE_HEARTBEAT_ERROR: 5 = 5;

export type PARENT_MESSAGE_ERROR =
  ...
  | typeof PARENT_MESSAGE_HEARTBEAT_ERROR;
```

SimenB replied:
That makes sense to me

jeysal replied:
Is this actually worth it though? It's not a message that is actually ever sent is it? It's just to  set a property on an error that can further identify the error but that I don't think we really need?
I think declaring messages that are not actually messages ever being passed between parent and child is confusing

jeysal commented:
@ayshiff isn't the problem you described already integration testing level? I would imagine the first tests for `processChild` and `ChildProcessWorker` as I roughly described them earlier wouldn't test such a real use case but only test that the interval is scheduled correctly in the `processChild` and handled correctly in the `ChildProcessWorker`

ayshiff commented:
@jeysal I come to ask for help if you have time 👍
I am trying to correct the tests that do not pass.
As Simen said, I'm exposing a factory function (`create`) for `jest-worker` instead of the class directly.

Inside the tests of `jest-haste-map` I mock `jest-worker` like this:

```js
jest.mock('jest-worker', () => ({
  create: jest.fn(worker => {
    mockWorker = jest.fn((...args) => require(worker).worker(...args));
    mockEnd = jest.fn();

    return {
      end: mockEnd,
      worker: mockWorker,
    };
  }),
}));
```

But in the test `'distributes work across workers'`, the `jestWorker.create.mock.calls` has a length of 5 instead of 1 and `mockWorker.mock.calls` has a length of 1 instead of 5 and it contains only the last worker instead of the five.

jeysal commented:
Hmm that's odd, I thought maybe it would have to be `jest.mock('jest-worker', () => ({ default: class Worker { create = jest.fn(...)}}))` because jest-worker is a default export, but if that's the problem I'd expect it to fail with 'create is not a function' ... @SimenB ideas?
@ayshiff can you perhaps push any test changes you made so that we can check it out

ayshiff commented:
> Hmm that's odd, I thought maybe it would have to be `jest.mock('jest-worker', () => ({ default: class Worker { create = jest.fn(...)}}))` because jest-worker is a default export, but if that's the problem I'd expect it to fail with 'create is not a function' ... @SimenB ideas?
> @ayshiff can you perhaps push any test changes you made so that we can check it out

@jeysal My test changes are already pushed inside `ChildProcessWorker.test.js` and `processChild.test.js`.

jeysal commented:
Haven't taken a look yet, but https://github.com/facebook/jest/issues/9054 just randomly popped into my inbox because someone commented, perhaps related?

jeysal commented:
Nope, #9054 was a false report

jeysal commented:
@ayshiff your `_getWorker` change in `jest-haste-map` was racy because of the way it accessed the shared `this._worker`. I pushed a fix for you. Let me know when you need anything else :)

ayshiff commented:
@jeysal  Now that I have been able to test that the behavior between process childs and workers was consistent (heartbeat logic inside `ChildProcessWorker.test.js` and `processChild.test.js`) I would like to be able to:

1. Correct failed tests

I suspect that some `setTimeout` are not well cleared.
When I run `yarn jest`, some tests fail with the heartbeat timeout message `Test worker was unresponsive for 5000 milliseconds. There was no inspector connected so we were unable to capture stack frames before it was terminated.` whereas when I run the test individually the problem does not appear.

Also, I tried to test if the inspector was properly initialized when we call the `Worker.create()` factory function and again when I run the test individually it works but when I run it with the others it tells me that the address on which it is launched is already in use: `Starting inspector on 127.0.0.1:9229 failed: address already in use`.

I imagine that the inspector is not properly closed between tests, but i can't find where.
(I tried to clear the timeout when we are into the other cases of the switch statement inside the `_onMessage` child message handler but without much success.)

2. Process error with the stack trace inside `monitorHeartbeatTimeout`

3. Add e2e tests for infinite loops detection

Thank you again for taking time to help me 👍

jeysal commented:
> 2\. Add e2e tests for infinite loops detection

I imagine this could be as straightforward as creating a test that does `while(true);`, running Jest on it and checking that it errors with the stack trace and everything.

> * Process error with the stack trace inside `monitorHeartbeatTimeout`
> *

I suppose we would do something similar to when receiving a CLIENT_ERROR, that is call `_onProcessEnd` (in this case we should probably check that there even is a current task running, which may in theory not be the case at the time the infinite loop runs, e.g. if the task scheduled a leftover timeout that then causes the infinite loop), then run the existing force exit logic?

---

Regarding the test timeouts: Could it be that some tests really just block that long? We have a very high test timeout set, and also consider that the e2e tests are basically doing `execSync('jest')`. It seems that at the moment the worker heartbeat timeout is fixed, but we may need to pass something more dynamic in, probably based on the `testTimeout` plus a certain value?
Perhaps it's not actually an issue with any clearing of timeouts. The fact that it passes when running one test could also be due to the limit not being exceeded when the machine is not under heavy load.
@SimenB this leads me to think: Don't we have a problem with dynamic `jest.setTimeout()`s, or even individual test timeouts, because we won't be able to tell the worker what the new maximum heartbeat interval is? :/

ayshiff commented:
I have the impression that the `jest-worker` is not instantiated when we have **only one** test file (e.g. when I create an e2e test with an infinite loop inside). The only time I see the the `jest-worker` instantiate (and therefore the `inspector` too) is when we have several tests running in parallel (e.g. `yarn jest`). 

My initial logic may not be good or I am not able to properly test the behavior of the jest-worker.

@jeysal Could you enlighten me on this point?

jeysal commented:
Oh yeah that is true. I believe we have something like 'force using workers' in our e2e test utils, maybe that could help

ayshiff commented:
Hello @jeysal!
It's been a while since I've had time to move forward on this PR.
I'm going to resume the development of the feature because I'd love to see it finished.

I'm going to get back to it and I'll surely come back with more questions to go over the blocking points.

The first point that I have trouble understanding is that the infinite loop can appear in a single test file while workers are used when you have several test files.
So I'm wondering if we need to make these changes in `jest-worker` (from what I understood).

Could we make a point of whether I'm going in the right direction and identify the things that still need to be done?

jeysal commented:
@ayshiff I'm sorry, I didn't look at Jest things for a while. Regarding the worker vs in-band thing - I think the assumption of all of us (not sure if it's ever been written or only implied) was that we would ignore the in-band case and only provide this feature if running in workers for now, because it's better than nothing and building it for in-band is difficult. So I do think this is going in the right direction.
Regarding TODOs, I believe the jest-worker code was essentially done? Did you manually verify that it kills workers stuck in an infinite loop? So we would only need to finish the testing. I think/hope other packages won't even need changes because when `jest-worker` terminates the process it passes the error message up as a reason and the user will already see it.

ayshiff commented:
> @ayshiff I'm sorry, I didn't look at Jest things for a while. Regarding the worker vs in-band thing - I think the assumption of all of us (not sure if it's ever been written or only implied) was that we would ignore the in-band case and only provide this feature if running in workers for now, because it's better than nothing and building it for in-band is difficult. So I do think this is going in the right direction.
> Regarding TODOs, I believe the jest-worker code was essentially done? Did you manually verify that it kills workers stuck in an infinite loop? So we would only need to finish the testing. I think/hope other packages won't even need changes because when `jest-worker` terminates the process it passes the error message up as a reason and the user will already see it.

No problem, it's also been a while since I've looked at this PR.

At the moment the heartbeat logic is tested and working:

- `ChildProcessWorker.test.js`: calls the onProcessEnd method when we have no heartbeat message from child during n-time
- `processChild.test.js`: should emit an heartbeat message every time interval

I tried to test the behavior of the workers in an e2e test that does `while(true)` (using `generateTestFilesToForceUsingWorkers`) but for some reason the workers are not killed.

I also tried to extend the timeout delay but some tests (by running all the tests) keep failing with the monitor heartbeat error.

I will surely need your help to find a solution to the worker not killed problem.

jeysal commented:
Okay so basically the individual pieces seem correct, but the whole thing altogether does not seem to work. I'll have to properly dive into this then (so far I've usually helped on specific limited aspects).
I think you didn't answer this in your comment (or I missed it): Did you ever try manually whether the feature works (because it might just be that something in our e2e test setup is weird and *usually* the it would already work).
(It's okay if you didn't try, I can do it, just asking because it might save time if you already know something)

ayshiff commented:
> Okay so basically the individual pieces seem correct, but the whole thing altogether does not seem to work. I'll have to properly dive into this then (so far I've usually helped on specific limited aspects).
> I think you didn't answer this in your comment (or I missed it): Did you ever try manually whether the feature works (because it might just be that something in our e2e test setup is weird and _usually_ the it would already work).
> (It's okay if you didn't try, I can do it, just asking because it might save time if you already know something)

Oh right, I misunderstood!
When I manually test, the feature works (the worker is killed when we don't receive a heartbeat message from the child process since n-time).
 What remains to be done is to display the stack trace retrieved by the inspector

jeysal commented:
Rebased and pushed some minor cleanup.
The error I'm getting on the e2e test is about the inspector being `open`ed when there is already an open inspector. However looking at the code related to the inspector/stacktrace bit, I'm not quite sure  I understand how it's supposed to work. Aren't you attaching the inspector to the worker parent, not the `processChild`ren?

ayshiff commented:
> Rebased and pushed some minor cleanup.
> The error I'm getting on the e2e test is about the inspector being `open`ed when there is already an open inspector. However looking at the code related to the inspector/stacktrace bit, I'm not quite sure I understand how it's supposed to work. Aren't you attaching the inspector to the worker parent, not the `processChild`ren?

Thanks for your help 👍
I am attaching the inspector to the worker parent. Should I attach it to `processChild` ?

jeysal commented:
I'm not very familiar with `inspector`, but I thought that's what you would want, because we're interested in the stack trace of the child when it is stopped in an infinite loop, right?

ayshiff commented:
Oh yes you are right.
My logic was probably wrong then, I'll try to move the inspector logic inside `processChild`.

I'll ping you when I'm done with it 👍
